core-bundle/src/Controller/ImagesController.php

```diff
@@ -0,0 +1,65 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Controller;
+
+use Contao\CoreBundle\Image\ImageFactoryInterface;
+use Contao\Image\DeferredImageInterface;
+use Contao\Image\DeferredResizerInterface;
+use Contao\Image\ResizerInterface;
+use Symfony\Component\HttpFoundation\BinaryFileResponse;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
+
+class ImagesController
+{
+    /**
+     * @var ImageFactoryInterface
+     */
+    private $imageFactory;
+
+    /**
+     * @var ResizerInterface
+     */
+    private $resizer;
+
+    /**
+     * @var string
+     */
+    private $targetDir;
+
+    public function __construct(ImageFactoryInterface $imageFactory, ResizerInterface $resizer, string $targetDir)
+    {
+        $this->imageFactory = $imageFactory;
+        $this->resizer = $resizer;
+        $this->targetDir = $targetDir;
+    }
+
+    /**
+     * The route is registered dynamically in the Contao\CoreBundle\Routing\ImagesLoader class.
+     */
+    public function __invoke(string $path): Response
+    {
+        try {
+            $image = $this->imageFactory->create($this->targetDir.'/'.$path);
+            $resizer = $this->resizer;
+
+            if ($image instanceof DeferredImageInterface && $resizer instanceof DeferredResizerInterface) {
+                $resizer->resizeDeferredImage($image);
+            }
+        } catch (\Exception $exception) {
+            throw new NotFoundHttpException($exception->getMessage(), $exception);
+        }
+
+        return new BinaryFileResponse($image->getPath());
```

Comment:
I think we should add caching headers here ;) We should cache it for a year. I also thought about adding an `ETag` so that after this year we can save bandwidth if the image still hasn't changed. However, I think that there's no real benefit. Calculating the hash for every image is too expensive, I prefer a re-download after a year. Also, the `BinaryFileResponse` is `public` by default which would then fill our reverse proxy. I don't think that makes a lot of sense because the images are created and thus cached anyway. So imho this should look like this:

```php
return new BinaryFileResponse($image->getPath(), 200, ['Cache-Control' => 'private, maxage=31536000'], false);
```

Reply:
I’m not sure about the `private` part. If someone doesn’t use our reverse proxy, public caching might be desired (e.g. on a CDN).

How about configuring our reverse proxy to never cache `image/*` responses?

Reply:
Setting it to private would also differ from the image responses that are handled by the web server directly: https://github.com/contao/contao/blob/4118d8fac40a4b51fbf85f91caf6ce92e02a2322/manager-bundle/src/Resources/skeleton/web/.htaccess#L8

Reply:
Cache headers added in 12761e48b6ee4e9a468d239169562e957de3646c

Reply:
Hmm, I don't think it would be different than our existing `.htaccess`. We don't add any `public` directive so I think it's treated as being `private`. But in any case, I agree we need to find the best solution here. So we have the following 2 use cases:

1. No special proxy setup = Symfony HttpCache. If we make the response `public`, it would - as of today - duplicate the images on the server. I think we can agree that there's no point in doing so. Now either we configure our `HttpCache` instance to ignore `Content-Type: image/*` or `/assets/images` responses (I'd prefer the path based one because it's more specific to only our thumbnails) or we make sure we do not send the `public` cache-control directive. The latter would be easier I think (we could add a configuration like `contao.image.controller_cache_control: max-age=1234, private` with the default value for the Symfony proxy). Having HttpCache ignoring certain responses will be tricky because we'd have to override the `fetch()` method because ignoring certain responses is not part of the design (yet): https://github.com/symfony/symfony/blob/4.3/src/Symfony/Component/HttpKernel/HttpCache/HttpCache.php#L452

2. Special proxy setup = e.g. Varnish: It might make sense to also store the generated images in Varnish or some CDN for performance reasons. However, we now also have some maintenance task which is called "clear image and scripts assets". Executing this would have no effect as it would clear files on the server but not in the proxy. If we wanted to support that, consequently we should add cache tags to those files (which would now work for the thumbnails but not for css and js files...). So that needs some special knowledge anyway. And for Varnish I could easily force the `/assets/images` requests to be public if I wanted to store them.

So not sure here really. I think it's better to make them private by default because that's what we need for our HttpCache solution aka the 99% use case. For everybody else they can maybe either force the caching by using proxy configuration (possible in Varnish and probably also in some CDNs I suppose) or we may introduce a configuration parameter but I'm not sure if this is worth it?

Reply:
Ah, and for 1) there's also the possibility to not cache the files in `assets/images`, make the responses `public` and thus let HttpCache handle the file caching. But I don't consider this to be an alternative because

a) File path on the server does not match the URL = no direct static assets delivery by nginx or Apache.
b) Again, would need cache tags to support the maintenance purge operation.
c) Would kill all thumbnails during `cache:clear`, `composer install` and thus basically (at least in my case) every deployment.

I just ruled that out myself and did not even add it to the list. But I wanted to complete it so others can read the reasoning behind it.

Reply:
I agree. Private by default makes sense then.

> we may introduce a configuration parameter but I'm not sure if this is worth it?

If a project uses a different proxy they could use a listener to modify the cache headers very easily I think. So a configuration parameter is not needed IMO.

Reply:
I agree. Private by default makes sense then.

> we may introduce a configuration parameter but I'm not sure if this is worth it?

If a project uses a different proxy they could use a listener to modify the cache headers very easily I think. So a configuration parameter is not needed IMO.

Reply:
Done in ecfaff2b8ee70d6ee732961fec2587e2c185ea97.

Reply:
Yeah, we may just provide that configuration because it's easy to do and then we make it a "supported use case". Also, it would allow to adjust the caching duration to a month or so in case somebody needs that. Imho it would just be a fairly simple thing to introduce and potentially make somebody happy in the future :)

Reply:
@contao/developers WDYT?

Reply:
I don't know if this really has to be configurable.